dynamic/internal/fixup/properties.go

```diff
@@ -0,0 +1,157 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// package fixup applies fixes to a [info.Provider] to ensure that it can generate a valid
+// schema and that the schema can generate valid SDKs in all all languages.
+//
+// package fixup is still in development and may expose breaking changes in minor
+// versions.
+package fixup
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge/info"
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+)
+
+// Default applies the default set of fixups to p.
+//
+// The set of fixups applied may expand over time, but it should not effect providers that
+// correctly compile in all languages.
+func Default(p *info.Provider) error {
+	return errors.Join(
+		fixPropertyConflict(p),
+		fixMissingIDs(p),
+	)
+}
+
+func fixMissingIDs(p *info.Provider) error {
+	getIDType := func(r *info.Resource) tokens.Type {
+		s := r.Fields["id"]
+		if s == nil {
+			return ""
+		}
+		return s.Type
+	}
+	return walkResources(p, func(r tfbridge.Resource) error {
+		id, hasID := r.TF.Schema().GetOk("id")
+		ok := hasID &&
+			(id.Type() == shim.TypeString || getIDType(r.Schema) == "string") &&
+			id.Computed()
+		if !ok {
+			r.Schema.ComputeID = missingID
+		}
+		return nil
+	})
+}
+
+func missingID(context.Context, resource.PropertyMap) (resource.ID, error) {
+	return "missing ID", nil
+}
+
+func fixPropertyConflict(p *info.Provider) error {
+	return walkResources(p, func(r tfbridge.Resource) error {
+		var retError error
+		r.TF.Schema().Range(func(key string, value shim.Schema) bool {
+			if fix := badPropertyName(key); fix != nil {
+				if r.Schema.Fields == nil {
+					r.Schema.Fields = map[string]*info.Schema{}
+				}
+
+				s, ok := r.Schema.Fields[key]
+				if !ok {
+					s = &info.Schema{}
+				}
+
+				if s.Name == "" {
+					var err error
+					s.Name, err = fix(r.TF)
+					if err != nil {
+						retError = fmt.Errorf("%q: %w", key, err)
+					}
+				}
+
+				if !ok {
+					r.Schema.Fields[key] = s
+				}
+			}
+
+			return true
+		})
+		return retError
+	})
+}
+
+func badPropertyName(key string) func(shim.Resource) (string, error) {
+	switch key {
+	case "urn":
+		return newName(urnReplaceList)
+	default:
+		return nil
+	}
+}
+
+var urnReplaceList = []string{
```

Comment:
given that this is a list of 1 and it is all internal anyway, is there a reason not to just try the one string and fail if it doesn't work for now?

We can refactor in the future if we do decide to expand the list.

Reply:
Good point. I'll simplify to what we actually use.